File: data-analystics/iceberg/code/chapter5/ecosystem-integration/flink-integration/basic-setup.java
/*
 * Flink Iceberg 集成基础示例
 * 
 * 本示例演示如何在 Flink 中配置 Iceberg 并进行基本的表操作。
 */

import org.apache.flink.api.common.RuntimeExecutionMode;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.CoreOptions;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.api.TableResult;
import org.apache.flink.types.Row;

public class FlinkIcebergBasicSetup {

    public static void main(String[] args) throws Exception {
        // 1. 创建 Flink 执行环境
        Configuration flinkConfig = new Configuration();
        flinkConfig.set(CoreOptions.DEFAULT_PARALLELISM, 4);
        
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(flinkConfig);
        env.setRuntimeMode(RuntimeExecutionMode.BATCH);
        
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

        System.out.println("=== Flink Iceberg 集成基础配置 ===");
        
        // 2. 配置 Iceberg Catalog
        configureIcebergCatalog(tableEnv);

        // 3. 创建数据库
        tableEnv.executeSql("CREATE DATABASE IF NOT EXISTS default");
        tableEnv.executeSql("USE default");

        // 4. 创建 Iceberg 表
        createIcebergTables(tableEnv);

        // 5. 插入测试数据
        insertTestData(tableEnv);

        // 6. 查询数据
        queryData(tableEnv);

        // 7. 更新数据
        updateData(tableEnv);

        // 8. 删除数据
        deleteData(tableEnv);

        // 9. 查看表元数据
        showTableMetadata(tableEnv);

        // 10. 时间旅行查询
        timeTravelQuery(tableEnv);

        // 11. 清理资源
        cleanup(tableEnv);

        System.out.println("Flink Iceberg 基础配置示例完成!");
    }

    private static void configureIcebergCatalog(StreamTableEnvironment tableEnv) {
        System.out.println("配置 Iceberg Catalog...");
        
        // 配置 Iceberg Catalog
        tableEnv.executeSql("CREATE CATALOG iceberg_catalog WITH (" +
                "'type'='iceberg'," +
                "'catalog-type'='hadoop'," +
                "'warehouse'='hdfs://localhost:9000/iceberg/warehouse'" +
                ")");
        
        tableEnv.executeSql("USE CATALOG iceberg_catalog");
    }

    private static void createIcebergTables(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("创建 Iceberg 表...");
        
        // 创建用户表
        tableEnv.executeSql("CREATE TABLE IF NOT EXISTS default.flink_users (" +
                "id BIGINT," +
                "name STRING," +
                "email STRING," +
                "age INT," +
                "created_at TIMESTAMP(3)" +
                ") WITH (" +
                "'write.format.default'='parquet'," +
                "'write.target-file-size-bytes'='268435456'," +
                "'write.parquet.compression-codec'='zstd'" +
                ")");
        
        // 创建订单表
        tableEnv.executeSql("CREATE TABLE IF NOT EXISTS default.flink_orders (" +
                "order_id BIGINT," +
                "user_id BIGINT," +
                "product_name STRING," +
                "amount DECIMAL(10,2)," +
                "order_date DATE," +
                "status STRING" +
                ") WITH (" +
                "'write.format.default'='parquet'," +
                "'write.target-file-size-bytes'='268435456'" +
                ")");
    }

    private static void insertTestData(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("插入测试数据...");
        
        // 插入用户数据
        tableEnv.executeSql("INSERT INTO default.flink_users VALUES " +
                "(1, '张三', 'zhangsan@example.com', 25, TIMESTAMP '2023-01-15 10:30:00'), " +
                "(2, '李四', 'lisi@example.com', 30, TIMESTAMP '2023-01-16 14:45:00'), " +
                "(3, '王五', 'wangwu@example.com', 28, TIMESTAMP '2023-01-17 09:15:00'), " +
                "(4, '赵六', 'zhaoliu@example.com', 35, TIMESTAMP '2023-01-18 16:20:00')").await();
        
        // 插入订单数据
        tableEnv.executeSql("INSERT INTO default.flink_orders VALUES " +
                "(1001, 1, '笔记本电脑', 5999.99, DATE '2023-01-20', 'completed'), " +
                "(1002, 2, '手机', 2999.99, DATE '2023-01-21', 'completed'), " +
                "(1003, 1, '鼠标', 99.99, DATE '2023-01-22', 'pending'), " +
                "(1004, 3, '键盘', 199.99, DATE '2023-01-23', 'shipped')").await();
        
        System.out.println("测试数据插入完成");
    }

    private static void queryData(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("查询所有用户数据:");
        TableResult userResult = tableEnv.executeSql("SELECT * FROM default.flink_users");
        userResult.print();
        
        System.out.println("查询所有订单数据:");
        TableResult orderResult = tableEnv.executeSql("SELECT * FROM default.flink_orders");
        orderResult.print();
        
        System.out.println("关联查询用户和订单:");
        TableResult joinResult = tableEnv.executeSql(
            "SELECT u.name, u.email, o.product_name, o.amount, o.status " +
            "FROM default.flink_users u " +
            "JOIN default.flink_orders o ON u.id = o.user_id"
        );
        joinResult.print();
    }

    private static void updateData(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("更新用户数据...");
        
        tableEnv.executeSql(
            "UPDATE default.flink_users " +
            "SET email = 'zhangsan_updated@example.com' " +
            "WHERE id = 1"
        ).await();
        
        System.out.println("更新后的用户数据:");
        TableResult result = tableEnv.executeSql("SELECT * FROM default.flink_users WHERE id = 1");
        result.print();
    }

    private static void deleteData(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("删除订单数据...");
        
        tableEnv.executeSql(
            "DELETE FROM default.flink_orders " +
            "WHERE order_id = 1004"
        ).await();
        
        System.out.println("删除后的订单数据:");
        TableResult result = tableEnv.executeSql("SELECT * FROM default.flink_orders");
        result.print();
    }

    private static void showTableMetadata(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("查看用户表的历史快照:");
        TableResult historyResult = tableEnv.executeSql("SELECT * FROM default.flink_users.history");
        historyResult.print();
        
        System.out.println("查看订单表的分区信息:");
        TableResult partitionsResult = tableEnv.executeSql("SELECT * FROM default.flink_orders.partitions");
        partitionsResult.print();
        
        System.out.println("查看用户表的当前快照:");
        TableResult snapshotsResult = tableEnv.executeSql("SELECT * FROM default.flink_users.snapshots");
        snapshotsResult.print();
    }

    private static void timeTravelQuery(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("时间旅行查询...");
        
        // 获取最早的快照ID
        TableResult snapshotResult = tableEnv.executeSql(
            "SELECT snapshot_id FROM default.flink_users.history ORDER BY committed_at ASC LIMIT 1"
        );
        
        // 注意：在实际应用中，我们需要从结果中提取快照ID
        // 这里为了简化示例，我们直接使用AS OF语法演示
        System.out.println("查询表的早期状态:");
        TableResult timeTravelResult = tableEnv.executeSql(
            "SELECT * FROM default.flink_users /*+ OPTIONS('scan.snapshot-id'='1') */"
        );
        // 注意：上面的注释语法可能因Flink版本而异，这里仅作示意
        
        System.out.println("当前表数据:");
        TableResult currentResult = tableEnv.executeSql("SELECT * FROM default.flink_users");
        currentResult.print();
    }

    private static void cleanup(StreamTableEnvironment tableEnv) throws Exception {
        System.out.println("清理测试表...");
        tableEnv.executeSql("DROP TABLE IF EXISTS default.flink_users");
        tableEnv.executeSql("DROP TABLE IF EXISTS default.flink_orders");
    }
}

File: data-analystics/iceberg/code/chapter5/ecosystem-integration/spark-integration/basic-setup.scala
/*
 * Spark Iceberg 集成基础示例
 * 
 * 本示例演示如何在 Spark 中配置 Iceberg 并进行基本的表操作。
 */

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

object SparkIcebergBasicSetup {
  def main(args: Array[String]): Unit = {
    // 1. 创建 Spark Session 并配置 Iceberg
    val spark = SparkSession.builder()
      .appName("Spark Iceberg Basic Setup")
      .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
      .config("spark.sql.catalog.spark_catalog", "org.apache.iceberg.spark.SparkSessionCatalog")
      .config("spark.sql.catalog.spark_catalog.type", "hadoop")
      .config("spark.sql.catalog.spark_catalog.warehouse", "hdfs://localhost:9000/iceberg/warehouse")
      .getOrCreate()

    import spark.implicits._

    println("=== Spark Iceberg 集成基础配置 ===")
    
    // 2. 创建数据库
    spark.sql("CREATE DATABASE IF NOT EXISTS default")
    spark.sql("USE default")
    
    // 3. 创建 Iceberg 表
    println("创建 Iceberg 表...")
    spark.sql("""
      CREATE TABLE IF NOT EXISTS users (
        id BIGINT,
        name STRING,
        email STRING,
        age INT,
        created_at TIMESTAMP
      )
      USING iceberg
      PARTITIONED BY (days(created_at))
      TBLPROPERTIES (
        'write.format.default'='parquet',
        'write.target-file-size-bytes'='536870912'
      )
    """)
    
    // 4. 插入测试数据
    println("插入测试数据...")
    val userData = Seq(
      (1L, "张三", "zhangsan@example.com", 25, java.sql.Timestamp.valueOf("2023-01-15 10:30:00")),
      (2L, "李四", "lisi@example.com", 30, java.sql.Timestamp.valueOf("2023-01-16 14:45:00")),
      (3L, "王五", "wangwu@example.com", 28, java.sql.Timestamp.valueOf("2023-01-17 09:15:00")),
      (4L, "赵六", "zhaoliu@example.com", 35, java.sql.Timestamp.valueOf("2023-01-18 16:20:00"))
    ).toDF("id", "name", "email", "age", "created_at")
    
    userData.writeTo("users").append()
    
    // 5. 查询数据
    println("查询所有用户数据:")
    spark.table("users").show(false)
    
    // 6. 条件查询
    println("查询年龄大于25的用户:")
    spark.sql("SELECT * FROM users WHERE age > 25").show(false)
    
    // 7. 聚合查询
    println("用户年龄统计:")
    spark.sql("""
      SELECT 
        COUNT(*) as total_users,
        AVG(age) as avg_age,
        MIN(age) as min_age,
        MAX(age) as max_age
      FROM users
    """).show()
    
    // 8. 查看表的元数据信息
    println("查看表的历史快照:")
    spark.sql("SELECT * FROM users.history").show(false)
    
    println("查看表的分区信息:")
    spark.sql("SELECT * FROM users.partitions").show(false)
    
    // 9. 更新数据
    println("更新用户数据...")
    spark.sql("""
      UPDATE users 
      SET email = 'zhangsan_new@example.com' 
      WHERE id = 1
    """)
    
    println("更新后的用户数据:")
    spark.sql("SELECT * FROM users WHERE id = 1").show(false)
    
    // 10. 删除数据
    println("删除用户数据...")
    spark.sql("DELETE FROM users WHERE id = 4")
    
    println("删除后的用户数据:")
    spark.table("users").show(false)
    
    // 11. 查看表的当前快照
    println("当前表快照信息:")
    spark.sql("SELECT * FROM users.snapshots").show(false)
    
    // 12. 时间旅行查询
    println("时间旅行查询 - 查看表的早期状态:")
    val earliestSnapshot = spark.sql("SELECT snapshot_id FROM users.history ORDER BY committed_at ASC LIMIT 1")
      .collect()(0).getAs[Long]("snapshot_id")
    
    spark.sql(s"SELECT * FROM users VERSION AS OF $earliestSnapshot").show(false)
    
    // 13. 创建分支
    println("创建分支...")
    spark.sql("ALTER TABLE users CREATE BRANCH test_branch")
    
    // 14. 在分支上插入数据
    println("在分支上插入数据...")
    val branchData = Seq(
      (5L, "孙七", "sunqi@example.com", 22, java.sql.Timestamp.valueOf("2023-01-19 11:30:00"))
    ).toDF("id", "name", "email", "age", "created_at")
    
    branchData.write.option("branch", "test_branch").mode("append").saveAsTable("users")
    
    println("主分支数据:")
    spark.table("users").show(false)
    
    println("test_branch 分支数据:")
    spark.read.option("branch", "test_branch").table("users").show(false)
    
    // 15. 合并分支
    println("合并分支...")
    spark.sql("CALL system.merge_branches(table => 'users', branch => 'test_branch')")
    
    println("合并后的数据:")
    spark.table("users").show(false)
    
    // 16. 查看表属性
    println("表属性信息:")
    spark.sql("SHOW TBLPROPERTIES users").show(false)
    
    // 17. 修改表属性
    println("修改表属性...")
    spark.sql("ALTER TABLE users SET TBLPROPERTIES ('comments' = '用户信息表')")
    
    println("修改后的表属性:")
    spark.sql("SHOW TBLPROPERTIES users").show(false)
    
    // 18. 清理资源
    println("清理测试表...")
    spark.sql("DROP TABLE IF EXISTS users")
    
    spark.stop()
    println("Spark Iceberg 基础配置示例完成!")
  }
}